<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="robot">

    <!-- <xacro:property name="pi" value="3.14" /> -->

    <link name="base_link" > 
        <visual>
            <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
            <geometry>
              <mesh filename="file://$(find robot_bringup)/robot_mesh/base_link.dae"/> 
            </geometry>
          </visual>
          <collision>
            <origin rpy="${pi/2} 0 0" xyz="0 0 0"/>
            <geometry>
              <mesh filename="file://$(find robot_bringup)/robot_mesh/base_link.dae"/> 
            </geometry>
          </collision>
          <inertial>
            <origin xyz="0.15 0.15 .2" rpy="0 0 0"/>
            <mass value="10000"/>
            <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
          </inertial>
    </link >

    <link name="link1">
      <visual>
        <origin rpy="${pi/2} 0 ${pi/2}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link1.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="${pi/2} 0 ${pi/2}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link1.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link2" >
      <visual>
        <origin rpy="-${pi/2} 0 ${pi}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link2.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="-${pi/2} 0 ${pi}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link2.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link3" >
      <visual>
        <origin rpy="${pi/2} 0 ${pi}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link3.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="${pi/2} 0 ${pi}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link3.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link4_artificial" >
      <inertial>
        <mass value="10"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link4" >
      <visual>
        <origin rpy="${pi} ${pi/2} 0" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="${pi} ${pi/2} 0" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link5" >
      <inertial>
        <mass value="10"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link4_1" >
      <visual>
        <origin rpy="${pi} ${pi/2} 0.261" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4_1.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="${pi} ${pi/2} 0.261" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4_1.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >
    
    <link name="link4_2" >
      <visual>
        <origin rpy="${pi} ${pi/2} ${pi/2}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4_2.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="${pi} ${pi/2} ${pi/2}" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4_2.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link4_3_artificial" >
      <inertial>
        <mass value="10"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >

    <link name="link4_3" >
      <visual>
        <origin rpy="0 0 0.12" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4_3.dae"/> 
        </geometry>
      </visual>
      <collision>
        <origin rpy="0 0 0.12" xyz="0 0 0"/>
        <geometry>
          <mesh filename="file://$(find robot_bringup)/robot_mesh/link4_3.dae"/> 
        </geometry>
      </collision>
      <inertial>
        <mass value="1000"/>
        <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
      </inertial>
    </link >
  

    <!-- Joints Declaration -->
    <joint name="joint1" type="continuous">
      <parent link="base_link"/>
      <child link="link1"/>
      <axis xyz="0 0 1" />
      <origin xyz="0 0 0.2645" rpy="0 0 0" />
    </joint>
  
    <joint name="joint2" type="continuous">
      <parent link="link1"/>
      <child link="link2"/>
      <axis xyz="0 0 1" />
      <origin xyz="0.247 0.16802 0.332" rpy="${pi/2} 0 0" />
    </joint>
  
    <joint name="joint3" type="continuous">
      <parent link="link2"/>
      <child link="link3"/>
      <axis xyz="0 0 1" />
      <origin xyz="0 1.100 0" rpy="0 0 0" />
    </joint>

    <joint name="joint4_artificial" type="continuous">
      <mimic joint="joint2" multiplier="-1" offset="0" />
      <parent link="link3"/>
      <child link="link4_artificial"/>
      <axis xyz="0 0 1" />
      <origin xyz="0.97283 0 0.1025" rpy="0 0 0" />
    </joint>

    <joint name="joint4" type="continuous">
      <mimic joint="joint3" multiplier="-1" offset="0" />
      <parent link="link4_artificial"/>
      <child link="link4"/>
      <axis xyz="0 0 1" />
      <origin xyz="0 0 0" rpy="0 0 0" />
    </joint>

    <joint name="joint5" type="continuous">
      <parent link="link4"/>
      <child link="link5"/>
      <axis xyz="0 0 1" />
      <origin xyz="0.17227 -0.18953 0" rpy="${pi/2} 0 0" />
    </joint>

    <joint name="joint4_1" type="continuous">
      <mimic joint="joint2" multiplier="-1" offset="0" />
      <parent link="link1"/>
      <child link="link4_1"/>
      <axis xyz="0 0 1" />
      <origin xyz="-0.08421 0.06192 0.37687" rpy="-${pi/2} 0.02 0" />
    </joint>

    <joint name="joint4_2" type="continuous">
      <mimic joint="joint2" multiplier="1" offset="0" />
      <parent link="link4_1"/>
      <child link="link4_2"/>
      <axis xyz="0 0 1" />
      <origin xyz="0.010 -1.100 -0.0245" rpy="0 0 -0.15" />
    </joint>

    <joint name="joint4_3_artificial" type="continuous">
      <mimic joint="joint2" multiplier="-1" offset="0" />
      <parent link="link4_2"/>
      <child link="link4_3_artificial"/>
      <axis xyz="0 0 1" />
      <origin xyz="0.58329 0 -0.070" rpy="0 0 0" />
    </joint>

    <joint name="joint4_3" type="continuous">
      <mimic joint="joint3" multiplier="-1" offset="0" />
      <parent link="link4_3_artificial"/>
      <child link="link4_3"/>
      <axis xyz="0 0 1" />
      <origin xyz="0 0 0" rpy="0 0 0" />
    </joint>

    <!-- Transmission -->
    <gazebo>
      <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
        <robotNamespace>/robot</robotNamespace>
        <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
      </plugin>
    </gazebo>

    <transmission name="tran1">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$base_link_motor">
        <mechanicalReduction>1</mechanicalReduction>
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </actuator>
      <joint name="joint1">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
    </transmission>

    <transmission name="tran2">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$link2_motor">
        <mechanicalReduction>1</mechanicalReduction>
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </actuator>
      <joint name="joint2">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
    </transmission>

    <transmission name="tran3">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$link3_motor">
        <mechanicalReduction>1</mechanicalReduction>
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </actuator>
      <joint name="joint3">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
    </transmission>



  </robot>